fileMonitor/fileMon.go

```diff
@@ -55,28 +56,28 @@ func New(cfg *Cfg) (*FileMon, error) {
 	}
 
 	fm := &FileMon{
-		cfg:        *cfg,
-		eventTable: make(map[string]bool),
-		cmdCh:      make(chan cmd),
-		eventCh:    make(chan []Event, cfg.EventBufSize),
+		cfg:       *cfg,
+		fileTable: make(map[string]bool),
+		stopCh:    make(chan struct{}),
+		eventCh:   make(chan []Event, cfg.EventBufSize),
 	}
 
-	go fileMon(fm)
-
 	return fm, nil
 }
 
 func (fm *FileMon) Add(file string) {
 	fm.mu.Lock()
-	fm.eventTable[file] = true
+	fm.fileTable[file] = true
+	if !fm.running {
+		fm.running = true
+		go fileMon(fm)
```

Comment:
I would have used message-passing instead of shared-mem (`fm.mu`) to inject files into the `fileTable` map. That way, you would have saved most of the lock/unlock calls over `fm.mu` -- you would just need one new channel to interact with the `fileMon` goroutine). Your implementation should work, too; it's just that I find it trickier to follow with the set of (potentially paralleled) `fm.mu.Lock()` calls. Pls go ahead and merge if you don't see a race condition here (I don't see it, but you might want to double-check with the [race detector](https://go.dev/doc/articles/race_detector)).

Reply:
Thanks, that's a valid point. The shared-mem approach on fileTable and associated lock was there from before though, not changing it in this PR.